src/slurm.c

```diff
@@ -0,0 +1,584 @@
+      {.derive = stats_resp->jobs_started},
+      {.derive = stats_resp->jobs_completed},
+      {.derive = stats_resp->jobs_canceled},
+      {.derive = stats_resp->jobs_failed},
+  };
+  slurm_submit("slurm_stats", "slurm_stats_jobs", NULL, jobs_values,
+               STATIC_ARRAY_SIZE(jobs_values));
+
+  value_t bf_values[] = {
+      {.derive = stats_resp->bf_backfilled_jobs},
+      {.derive = stats_resp->bf_backfilled_pack_jobs},
+      {.counter = stats_resp->bf_cycle_counter},
+      {.gauge = stats_resp->bf_cycle_last},
+      {.counter = stats_resp->bf_cycle_max},
+      {.counter = stats_resp->bf_cycle_sum},
+      {.gauge = stats_resp->bf_last_depth},
+      {.gauge = stats_resp->bf_last_depth_try},
+      {.counter = stats_resp->bf_depth_sum},
+      {.counter = stats_resp->bf_depth_try_sum},
+      {.gauge = stats_resp->bf_queue_len},
+      {.counter = stats_resp->bf_queue_len_sum},
+  };
+  slurm_submit("slurm_stats", "slurm_stats_backfill", NULL, bf_values,
+               STATIC_ARRAY_SIZE(bf_values));
+}
+
+static int slurm_read(void) {
+  job_info_msg_t *job_buffer_ptr = NULL;
+  job_info_t *job_ptr;
+  partition_info_msg_t *part_buffer_ptr = NULL;
+  partition_info_t *part_ptr;
+  partition_state_t *partition_states;
+  partition_state_t *partition_state;
+  node_info_msg_t *node_buffer_ptr = NULL;
+  node_info_t *node_ptr;
+  stats_info_response_msg_t *stats_resp;
+  stats_info_request_msg_t stats_req;
+
+  if (slurm_load_jobs((time_t)NULL, &job_buffer_ptr, SHOW_ALL)) {
+    ERROR("slurm_load_jobs error");
+    return -1;
+  }
+
+  if (slurm_load_node((time_t)NULL, &node_buffer_ptr, SHOW_ALL)) {
+    slurm_free_job_info_msg(job_buffer_ptr);
+    ERROR("slurm_load_node error");
+    return -1;
+  }
+
+  if (slurm_load_partitions((time_t)NULL, &part_buffer_ptr, 0)) {
+    slurm_free_job_info_msg(job_buffer_ptr);
+    slurm_free_node_info_msg(node_buffer_ptr);
+    ERROR("slurm_load_partitions error");
+    return -1;
+  }
+
+  stats_req.command_id = STAT_COMMAND_GET;
+  if (slurm_get_statistics(&stats_resp, &stats_req)) {
+    slurm_free_job_info_msg(job_buffer_ptr);
+    slurm_free_node_info_msg(node_buffer_ptr);
+    slurm_free_partition_info_msg(part_buffer_ptr);
+    ERROR("slurm_get_statistics error");
+  }
+
+  /* SLURM APIs provide *non-relational* data about nodes, partitions and jobs.
+   * We allocate a data structure that relates all three together, and the
+   * following
+   * two for loops fill this data structure. The data structure is an array
+   * of partition_state_t that holds job and node states. */
+  uint32_t num_partitions = part_buffer_ptr->record_count;
+  partition_states =
+      alloc_partition_states(num_partitions, part_buffer_ptr->partition_array);
+  if (!partition_states) {
+    slurm_free_job_info_msg(job_buffer_ptr);
+    slurm_free_node_info_msg(node_buffer_ptr);
+    slurm_free_partition_info_msg(part_buffer_ptr);
+    ERROR("alloc_partition_states");
```

Comment:
Suggestion: add PLUGIN_NAME to all ERRORs like:
`ERROR(PLUGIN_NAME ": alloc_partition_states error.");`